### PR TITLE
Feature/243 calculate sponsored percentage

### PR DESCRIPTION
**Pull request by NicolasDontigny**
Implemented the calculations for the sponsored progress of a site.

I had to update models, serializers, and so I squashed the migrations to one initial script again.
Sponsor progress is now calculated dynamically according to the number of species of a site and of its batches, it is not a field that should be stored in the model/DB

### CODE REVIEW

**Comment by Samuel-Therrien-Beslogic** [comment]
Remember to link your PR and issue together. You can either use the "Development" menu on the left, or use a "closing keyword" in your description https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: canopeum_backend/canopeum_backend/models.py [hunk 1]
```diff
@@ -218,6 +234,10 @@ def add_supported_specie_by_id(self, pk: int):
         tree_type = Treetype.objects.get(pk=pk)
         return BatchSupportedSpecies.objects.create(tree_type=tree_type, batch=self)
 
+    def plant_count(self):
+        batch_species = BatchSpecies.objects.filter(batch=self)
+        return reduce(lambda x, y: x + y.quantity, batch_species, 0)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:239]: A reduce+add is just a sum ;)
```suggestion
        return sum(specie.quantity for specie in batch_species)
```

File: canopeum_backend/canopeum_backend/models.py [hunk 2]
```diff
@@ -156,6 +157,22 @@ class Site(models.Model):
     announcement = models.ForeignKey(Announcement, models.SET_NULL, blank=True, null=True)
     image = models.ForeignKey(Asset, models.SET_NULL, blank=True, null=True)
 
+    def get_plant_count(self):
+        site_species = Sitetreespecies.objects.filter(site=self)
+        return reduce(lambda x, y: x + y.quantity, site_species, 0)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:162]: A reduce+add is just a sum ;)
```suggestion
        return sum(specie.quantity for specie in site_species)
```

File: canopeum_backend/canopeum_backend/models.py [hunk 3]
```diff
@@ -156,6 +157,22 @@ class Site(models.Model):
...
+
+    def get_sponsor_progress(self):
+        total_plant_count = self.get_plant_count()
+        if total_plant_count == 0:
+            return 0
+
+        batches = Batch.objects.filter(site=self)
+        sponsored_plant_count = reduce(lambda x, y: x + y.plant_count(), batches, 0)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:170]: A reduce+add is just a sum ;)
```suggestion
        sponsored_plant_count = sum(batch.plant_count() for specie in batches)
```

**Review by Samuel-Therrien-Beslogic — CHANGES_REQUESTED** [review]
Your typing error should be fixed simply by adding the return type annotation on your model methods.

It would be nicer if the default dev DB actually came with various progressions (it can be randomized just like before, but on DB creation instead of every get)
image

File: canopeum_backend/canopeum_backend/models.py [hunk 4]
```diff
@@ -157,17 +156,17 @@ class Site(models.Model):
...
 
-    def get_sponsor_progress(self):
+    def get_sponsor_progress(self) -> float:
         total_plant_count = self.get_plant_count()
         if total_plant_count == 0:
             return 0
 
         batches = Batch.objects.filter(site=self)
-        sponsored_plant_count = reduce(lambda x, y: x + y.plant_count(), batches, 0)
+        sponsored_plant_count = sum(batch.plant_count() for specie in batches)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:169]: My bad, I mis-copy-pasted
```suggestion
        sponsored_plant_count = sum(batch.plant_count() for batch in batches)
```

File: canopeum_backend/canopeum_backend/models.py [hunk 5]
```diff
@@ -156,6 +156,22 @@ class Site(models.Model):
...
+        site_species = Sitetreespecies.objects.filter(site=self)
+        return sum(specie.quantity for specie in site_species)
+
+    def get_sponsor_progress(self) -> float:
+        total_plant_count = self.get_plant_count()
+        if total_plant_count == 0:
+            return 0
+
+        batches = Batch.objects.filter(site=self)
+        sponsored_plant_count = sum(batch.plant_count() for specie in batches)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:169]: My bad, I mis-copy-pasted
```suggestion
        sponsored_plant_count = sum(batch.plant_count() for batch in batches)
```

File: canopeum_backend/canopeum_backend/management/commands/initialize_database.py [hunk 6]
```diff
@@ -273,8 +273,8 @@ def create_batches_for_site(site):
     num_batches = random.randint(3, 8)
     for i in range(num_batches):
         number_of_seed = random.randint(50, 200)
-        plant_count = random.randint(0, number_of_seed)
-        survived_count = random.randint(0, plant_count)
+        survived_count = random.randint(100, 200)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/management/commands/initialize_database.py:276]: example
```suggestion
        survived_count = random.randint(100, 300)
```

File: canopeum_backend/canopeum_backend/models.py [hunk 7]
```diff
@@ -218,6 +233,10 @@ def add_supported_specie_by_id(self, pk: int):
         tree_type = Treetype.objects.get(pk=pk)
         return BatchSupportedSpecies.objects.create(tree_type=tree_type, batch=self)
 
+    def plant_count(self):
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:236]:
```suggestion
    def plant_count(self) -> int:
```

**Comment by NicolasDontigny** [comment]
> Your typing error should be fixed simply by adding the return type annotation on your model methods.
> 
> It would be nicer if the default dev DB actually came with various progressions (it can be randomized just like before, but on DB creation instead of every get) image

Added type annotations

**Review by Samuel-Therrien-Beslogic — APPROVED** [review]

**Pull request merged**